File: C++/pico-boilerplate/pages/add_sensor.cpp
//
// Created by Petr Jurásek on 08.04.2024.
//

#include "add_sensor.h"
#include "../sensors/SensorType.h"
#include "../config/setup.h"

namespace AddSensor{
    Display& display = Display::getInstance();
    pimoroni::ST7789& driver = display.getDriver();
    pimoroni::PicoGraphics_PenRGB332& graphics = display.getGraphics();
    Buttons& buttons = Buttons::getInstance();
    SensorManager& sensor_manager = SensorManager::getInstance();
}

using namespace AddSensor;

bool render_add_sensor(bool is_setup) {
    clear_fast();

    graphics.set_pen(Colors::WHITE);
    graphics.text("Add new sensor", pimoroni::Point{2, 0}, 236, 2);

#ifdef TEST_BUILD
    printf("Adding a new sensor\nEntering the select_sensor_type function\n");
#endif

    SensorType selected_sensor = select_sensor_type(is_setup);
    if (selected_sensor == SensorType::UNKNOWN) {
        return false;
    }

#ifdef TEST_BUILD
    printf("... Selected sensor type: %s\n", sensor_type_to_string(selected_sensor).c_str());
    printf("Entering the select_sensor_pin function\n");
#endif

    int selected_pin = select_sensor_pin(is_setup);
    if (selected_pin == -1) {
        return false;
    }

#ifdef TEST_BUILD
    printf("... Selected pin: %d\n", selected_pin);
    printf("Entering the select_sensor_name function\n");
#endif

    std::string sensor_name = select_sensor_name(is_setup);
    if (sensor_name.empty()) {
        return false;
    }

#ifdef TEST_BUILD
    printf("... Selected name: %s\n", sensor_name.c_str());
    printf("Entering the select_interval function\n");
#endif

    int selected_interval = select_interval(is_setup);
    if (selected_interval == -1) {
        return false;
    }

#ifdef TEST_BUILD
    printf("... Selected interval: %d\n", selected_interval);

    printf("All the data: \n");
    printf("Sensor type: %s\n Pin: %d\n Name: %s\n Interval: %d\n",
           sensor_type_to_string(selected_sensor).c_str(), selected_pin, sensor_name.c_str(), selected_interval);
    printf("Adding the sensor to the sensor manager\n");
#endif

    if (selected_sensor == SensorType::DHT11) {
        sensor_manager.addSensor(std::make_unique<DHT11>(DHT11(selected_pin,
                                                               sensor_name, selected_interval)));
    }

    return true;
}

SensorType select_sensor_type(bool is_setup) {
    std::vector<std::string> possibleSensors = sensorTypes;
    int sensor_list_length = possibleSensors.size();
    int selected_sensor_index = 0;

    sleep_ms(500);

    while (true){
        if (buttons.is_button_a_pressed()){
            selected_sensor_index = (selected_sensor_index + 1) % sensor_list_length;
        } else if (buttons.is_button_b_pressed()){
            if (selected_sensor_index == 0) { //C++ modulo operator doesn't work with neg numbers has to be done this way
                selected_sensor_index = sensor_list_length;
            }
            selected_sensor_index = selected_sensor_index - 1;
        } else if (buttons.is_button_x_pressed()){
#ifdef TEST_BUILD
            printf("Selected sensor: %s\n", possibleSensors[selected_sensor_index].c_str());
#endif

            if (is_setup) {
                help_interrupt();
            } else {
                return static_cast<SensorType>(selected_sensor_index);
            }
        } else if (buttons.is_button_y_pressed()){
#ifdef TEST_BUILD
            printf("User canceled the operation\n");
#endif

            if (is_setup) {
                return static_cast<SensorType>(selected_sensor_index);
            }

            return SensorType::UNKNOWN;
        }

        clear_fast();
        graphics.set_pen(Colors::WHITE);

        graphics.text("ADD SENSOR", Point{30, 8}, 200, 2);
        render_nav_arrows(2, Colors::WHITE);

        std::string text = "Select the sensor type: " +
                           sensor_type_to_string(static_cast<SensorType>(selected_sensor_index));
        graphics.text(text.c_str(), Point{2, 50}, 200, 2);

        graphics.set_pen(Colors::GREEN);
        // If the function is called from the setup, act as OK button, otherwise cancel button.
        if (is_setup) {
            graphics.text("OK", Point{DISPLAY_WIDTH - 30, DISPLAY_HEIGHT - 20}, 200, 2);
            graphics.set_pen(Colors::RED);
            graphics.text("HELP", Point{DISPLAY_WIDTH - 45, 20}, 200, 2);
        } else {
            graphics.text("OK", Point{DISPLAY_WIDTH - 30, 20}, 200, 2);
            graphics.set_pen(Colors::RED);
            graphics.text("CANCEL", Point{DISPLAY_WIDTH - 65, DISPLAY_HEIGHT - 20}, 200, 2);
        }
        graphics.set_pen(Colors::WHITE);

        driver.update(&graphics);
        sleep_ms(200);
    }
}

int select_sensor_pin(bool is_setup) {
    std::vector<int> pin_list = sensor_manager.getAvailablePins();
    int pin_list_length = pin_list.size();
    int selected_pin_index = 0;

    sleep_ms(200);

    while (true){
        if (buttons.is_button_a_pressed()) {
            selected_pin_index = (selected_pin_index + 1) % pin_list_length;
        } else if (buttons.is_button_b_pressed()) {
            if (selected_pin_index == 0) { //C++ modulo operator doesn't work with neg numbers has to be done this way
                selected_pin_index = pin_list_length;
            }
            selected_pin_index = selected_pin_index - 1;
        } else if (buttons.is_button_x_pressed()) {
            #ifdef TEST_BUILD
                printf("Selected pin: %d\n", pin_list[selected_pin_index]);
            #endif

            if (is_setup) {
                help_interrupt();
            } else {
                return pin_list[selected_pin_index];
            }
        } else if (buttons.is_button_y_pressed()) {
            #ifdef TEST_BUILD
                printf("User canceled the operation on selecting a pin.\n");
            #endif

            // If the function is called from the setup, act as OK button, otherwise cancel button.
            if (is_setup) {
                return pin_list[selected_pin_index];
            }

            return -1;
        }

        clear_fast();
        graphics.set_pen(Colors::WHITE);

        graphics.text("ADD SENSOR", Point{30, 8}, 200, 2);
        render_nav_arrows(2, Colors::WHITE);

        std::string text = "Select sensor pin: " + std::to_string(pin_list[selected_pin_index]);
        graphics.text(text.c_str(), Point{2, 60}, 200, 2);

        graphics.set_pen(Colors::GREEN);

        // If the function is called from the setup, act as OK button, otherwise cancel button.
        if (is_setup) {
            graphics.text("OK", Point{DISPLAY_WIDTH - 65, DISPLAY_HEIGHT - 20}, 200, 2);
            graphics.set_pen(Colors::RED);
            graphics.text("HELP", Point{DISPLAY_WIDTH - 30, 20}, 200, 2);
        } else {
            graphics.text("OK", Point{DISPLAY_WIDTH - 30, 20}, 200, 2);
            graphics.set_pen(Colors::RED);
            graphics.text("CANCEL", Point{DISPLAY_WIDTH - 65, DISPLAY_HEIGHT - 20}, 200, 2);
        }
        graphics.set_pen(Colors::WHITE);

        driver.update(&graphics);
        sleep_ms(150);
    }
}

std::string select_sensor_name(bool is_setup) {
    const char characters[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ_";
    const int characters_length = sizeof(characters) - 1;
    int current_index = 0;
    std::string name = "";




    while (true){
        sleep_ms(200);
        if (buttons.is_button_a_pressed()){
            current_index = (current_index + 1) % characters_length;
        } else if (buttons.is_button_b_pressed()){
            if (current_index == 0) {
                current_index = characters_length;
            }
            current_index = current_index - 1;
        } else if (buttons.is_button_x_pressed()){
            if (name.length() < 6) {
                name += (characters[current_index] == '_') ? ' ' : characters[current_index];
                current_index = 0;
            } else {
                return name;
            }
        } else if (buttons.is_button_y_pressed()){
            if (name.length() > 0) {
                name.pop_back();
            } else {
                if (is_setup) {
                    help_interrupt();
                } else {
                    return "";
                }
            }
        }

        clear_fast();
        graphics.set_pen(Colors::WHITE);

        graphics.text("ADD SENSOR", Point{30, 8}, 200, 2);
        render_nav_arrows(2, Colors::WHITE);

        graphics.text("Name of the sensor:", Point{2, 50}, 200, 2);
        graphics.text(name.c_str(), Point{2, 70}, 200, 2);

        auto length = graphics.measure_text(name.c_str()) + 2;
        if (name.length() < 6){
            graphics.set_pen(Colors::GRAY);
            graphics.text(std::string(1, characters[current_index]), Point{length, 70}, 200, 2);
        }

        graphics.set_pen(Colors::GREEN);
        if (name.length() == 6) {
            graphics.text("OK", Point{DISPLAY_WIDTH - 30, 20}, 200, 2);
        } else {
            graphics.text("ADD", Point{DISPLAY_WIDTH - 35, 20}, 200, 2);
        }

        graphics.set_pen(Colors::RED);
        if (name.length() == 0) {
            if (is_setup) {
                graphics.text("HELP", Point{DISPLAY_WIDTH - 45, DISPLAY_HEIGHT - 20}, 200, 2);
            } else {
                graphics.text("CANCEL", Point{DISPLAY_WIDTH - 65, DISPLAY_HEIGHT - 20}, 200, 2);
            }
        } else {
            graphics.text("DEL", Point{DISPLAY_WIDTH - 35, DISPLAY_HEIGHT - 20}, 200, 2);
        }

        driver.update(&graphics);
    }
}

int select_interval(bool is_setup) {
    int interval_time = 0;

    graphics.set_pen(Colors::WHITE);

    while (true){
        sleep_ms(200);
        if (buttons.is_button_a_pressed()) {
            interval_time = (interval_time + 1) % 61;
        } else if (buttons.is_button_b_pressed()) {
            interval_time = (interval_time - 1) % 61;
        } else if (buttons.is_button_x_pressed()) {
            if (is_setup) {
                help_interrupt();
            } else {
                return interval_time;
            }
        } else if (buttons.is_button_y_pressed()) {
            if (is_setup) {
                return interval_time;
            }

            return -1;
        }

        clear_fast();
        graphics.set_pen(Colors::WHITE);
        std::string text = "Select the interval: " + std::to_string(interval_time) + " min";
        graphics.text(text.c_str(), Point{2, 20}, 200, 2);

        if (is_setup) {

        }

        graphics.set_pen(Colors::GREEN);
        // If the function is called from the setup, act as OK button, otherwise cancel button.
        if (is_setup) {
            graphics.text("OK", Point{DISPLAY_WIDTH - 65, DISPLAY_HEIGHT - 20}, 200, 2);
            graphics.set_pen(Colors::RED);
            graphics.text("HELP", Point{DISPLAY_WIDTH - 30, 20}, 200, 2);
        } else {
            graphics.text("OK", Point{DISPLAY_WIDTH - 30, 20}, 200, 2);
            graphics.set_pen(Colors::RED);
            graphics.text("CANCEL", Point{DISPLAY_WIDTH - 65, DISPLAY_HEIGHT - 20}, 200, 2);
        }
        graphics.set_pen(Colors::WHITE);

        driver.update(&graphics);
    }
}

File: MicroPython/pages/add_sensor.py
from utime import sleep
from config.config import Colors, Display
from config.buttons import button_a, button_b, button_x, button_y
from sensors.sensor_manager import SensorManager, AvailableSensors
from graphics import page_elements
from sensors.dht11 import DHT11
from graphics.graphics import draw_next, draw_confirm, draw_cancel, draw_back, draw_help
from utils.misc_utils import help_interrupt


def render_add_sensor(is_initial=False):
    """
    Renders the page for adding a new sensor.

    :param is_initial: If this is called from the setup wizard to be set as a first sensor.

    :return: True if the sensor was added, False if the user cancelled the process.
    """

    display = Display()

    page_elements.clear_fast()

    display().set_pen(Colors.WHITE)
    display().text("Add new sensor", 2, 0, 236, 2)

    selected_sensor = select_sensor_type(is_initial)
    if selected_sensor is None:
        return False

    selected_pin = select_sensor_pin(is_initial)
    if selected_pin is None:
        return False

    selected_name = select_sensor_name(is_initial)
    if selected_name is "":
        return False

    selected_interval = select_interval(is_initial)
    if selected_interval is None:
        return False

    sensors = SensorManager()
    sensors.add_sensor(DHT11(selected_pin, selected_name, selected_interval))

    return True


def select_sensor_type(is_initial=False):
    """Selects the sensor type

    :param is_initial: If this is called from the setup wizard to be set as a first sensor.

    :return: The selected sensor type.
    """

    display = Display()

    sensor_list = AvailableSensors.list_sensors()
    sensor_list_length = len(sensor_list)
    selected_sensor = 0

    while True:
        sleep(.1)
        if button_a.is_held():
            selected_sensor = (selected_sensor + 1) % sensor_list_length

        if button_b.is_held():
            selected_sensor = (selected_sensor - 1) % sensor_list_length

        if button_x.read():
            return sensor_list[selected_sensor]

        if button_y.read():
            if is_initial:
                help_interrupt()
            else:
                return None

        page_elements.clear_fast()
        display().set_pen(Colors.WHITE)
        display().text(f"Select the sensor type: {sensor_list[selected_sensor]}", 2, 20, 170, 2)
        draw_next()

        if is_initial:
            draw_help()
        else:
            draw_cancel()
        display().update()


def select_sensor_pin(is_initial=False):
    """Selects the sensor pin

    :param is_initial: If this is called from the setup wizard to be set as a first sensor.

    :return: The selected sensor pin.
    """

    display = Display()

    pin_list = SensorManager().load_available_pins()
    pin_list_length = len(pin_list)
    selected_pin = 0

    while True:
        sleep(.1)
        if button_a.is_held():
            selected_pin = (selected_pin + 1) % pin_list_length

        if button_b.is_held():
            selected_pin = (selected_pin - 1) % pin_list_length

        if button_x.read():
            return int(pin_list[selected_pin][2])

        if button_y.read():
            if is_initial:
                help_interrupt()
            else:
                return None

        page_elements.clear_fast()
        display().set_pen(Colors.WHITE)
        display().text(f"Select the sensor pin: {pin_list[selected_pin]}", 2, 20, 180, 2)
        draw_next()

        if is_initial:
            draw_help()
        else:
            draw_cancel()
        display().update()


def select_sensor_name(is_initial=False):
    """Name the sensor. Max 6 character long name.

    :param is_initial: If this is called from the setup wizard to be set as a first sensor.

    :return: The name of the sensor.
    """

    display = Display()
    characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ_"
    len_characters = len(characters)
    name = ""
    current_index = 0

    while True:
        sleep(.1)
        if button_a.is_held():
            current_index = (current_index + 1) % len_characters
        elif button_b.is_held():
            current_index = (current_index - 1) % len_characters
        elif button_x.read():
            if len(name) < 6:
                name += (characters[current_index]) if characters[current_index] != "_" else " "
                current_index = 0
            else:
                return name.rstrip()

        if button_y.read():
            if len(name) > 0:
                name = name[:-1]
            elif is_initial:
                help_interrupt()
            else:
                return ""

        page_elements.clear_fast()
        display().set_pen(Colors.WHITE)
        display().text(f"Name of the sensor:", 2, 20, 180, 2)
        display().text(name, 2, 50, 236, 2)

        if len(name) < 6:
            draw_next()
        else:
            draw_confirm()

        if len(name) > 0:
            draw_back()
        elif is_initial:
            draw_help()
        else:
            draw_cancel()

        length = display().measure_text(name, 2) + 2
        if len(name) < 6:
            display().set_pen(Colors.GREY)
            display().text(characters[current_index], length, 50, 236, 2)

        display().update()


def select_interval(is_initial=False):
    """Selects the graph_strategy interval.

    :param is_initial: If this is called from the setup wizard to be set as a first sensor.

    :return: The selected graph_strategy interval.
    """

    display = Display()
    interval_time = 0

    display().set_pen(Colors.WHITE)

    while True:
        sleep(.1)
        if button_a.is_held():
            interval_time += 1

        if button_b.is_held():
            interval_time -= 1 if interval_time > 0 else 0

        if button_x.read():
            return interval_time * 60

        if button_y.read():
            if is_initial:
                help_interrupt()
            else:
                return None

        page_elements.clear_fast()
        display().set_pen(Colors.WHITE)
        display().text(f"Select the refresh interval of the sensor", 2, 20, 150, 2)
        display().text(f"Interval: {interval_time} m", 2, 80, 236, 2)
        draw_confirm()

        if is_initial:
            draw_help()
        else:
            draw_cancel()
        display().update()


if __name__ == "__main__":
    # select_sensor_name()
    # select_sensor_pin()
    # select_sensor_type()
    # select_interval()
    render_add_sensor(True)
